tests/testthat/helper-TealAppDriver.R

```diff
@@ -23,3 +23,51 @@ init_teal_app_driver <- function(...) {
     .package = "teal"
   )
 }
+
+ns_dataset <- function(prefix, suffix, dataset, extract = "singleextract") {
+  sprintf("%s-dataset_%s_%s-%s", prefix, dataset, extract, suffix)
+}
+
+test_plot_changes_no_errors <- function(app_driver, input_id, value, ws) {
+  test_object_no_changes(app_driver, input_id, value, ws, 'get_active_module_pws_output')
+}
+
+test_table_changes_no_errors <- function(app_driver, input_id, value, ws) {
+  test_object_no_changes(app_driver, input_id, value, ws, 'get_active_module_tws_output')
+}
+
+test_object_no_changes <- function(app_driver, input_id, value, ws, fun) {
+  object <- app_driver[[fun]]()(ws)
+  app_driver$set_active_module_input(input_id, value)
+  app_driver$expect_no_validation_error()
+  testthat::expect_false(identical(object, app_driver[[fun]]()(ws)))
+}
+
+test_plot_validation_error <- function(app_driver, input_id, value = character(0), message = NULL, validation = NULL) {
+  test_validation_error(app_driver, input_id, value, message, validation)
+}
+
+test_table_validation_error <- function(app_driver, input_id, value = character(0), message = NULL, validation = NULL,
+                                        tws = "table") {
+  test_validation_error(app_driver, input_id, value, message, validation)
+  testthat::expect_identical(app_driver$get_active_module_tws_output(tws), data.frame())
+}
+
+test_validation_error <- function(app_driver, input_id, value = character(0), message = NULL, validation = NULL) {
```

Comment:
I'm not sure. Trying to find relevant function name
```suggestion
test_set_input_validate <- function(app_driver, input_id, value = character(0), message = NULL, validation = NULL) {
```

```suggestion
expect_set_input_validate_error <- function(app_driver, input_id, value = character(0), message = NULL, validation = NULL) {
```